server/test/write_authzmodel.go

```diff
@@ -532,6 +532,117 @@ func WriteAuthorizationModelTest(t *testing.T, datastore storage.OpenFGADatastor
 			},
 			err: errors.InvalidAuthorizationModelInput(typesystem.InvalidRelationError("repo", "viewer")),
 		},
+		{
+			name: "Execute_Write_Fails_If_Recursive_Definition_For_A_Relation",
+			request: &openfgapb.WriteAuthorizationModelRequest{
+				StoreId: storeID,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "repo",
+						Relations: map[string]*openfgapb.Userset{
+							"viewer": {
+								Userset: &openfgapb.Userset_TupleToUserset{
+									TupleToUserset: &openfgapb.TupleToUserset{
+										Tupleset:        &openfgapb.ObjectRelation{Relation: "viewer"},
+										ComputedUserset: &openfgapb.ObjectRelation{Relation: "viewer"},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			err: errors.InvalidAuthorizationModelInput(typesystem.InvalidRelationError("repo", "viewer")),
+		},
+		{
+			name: "Execute_Write_Fails_If_Recursive_Definitions_For_A_Type",
+			request: &openfgapb.WriteAuthorizationModelRequest{
+				StoreId: storeID,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "repo",
+						Relations: map[string]*openfgapb.Userset{
+							"reader": {
+								Userset: &openfgapb.Userset_ComputedUserset{
+									ComputedUserset: &openfgapb.ObjectRelation{
+										Relation: "writer",
+									},
+								},
+							},
+							"writer": {
+								Userset: &openfgapb.Userset_ComputedUserset{
+									ComputedUserset: &openfgapb.ObjectRelation{
+										Relation: "reader",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			//TODO
```

Comment:
I think it's better to wait until the `graph` package is ready, add a `HasCycles` method, and use that

Reply:
Add it to the `typesystem` package 😄 . That's where this kind of stuff would best reside.

Reply:
`typesystem.HasCycles` doesn't make much sense to me. `graph.HasCycles` and `typeSystem.HasCycleInRelations` (which would simply delegate) makes more sense

Reply:
> Is the TODO to add a test for cycles

No, the test is there, the TODO is to add code that would make it pass 🙈 

But i'll leave it out of this PR for now. Later on we can add more tests according to the [document](https://oktawiki.atlassian.net/wiki/spaces/SDCSTL/pages/2708026620/Model+Validations) that Raghd wrote

Reply:
`WriteAuthorizationModel` is the RPC in question here, and we are adding validations that make sure we don't encounter cyclical definitions in the validation of the authz model provided in the request input. For this exact reason it makes the most sense to me to add this to the `typesystem.Validation(model *openfgapb.AuthorizationModel)` method. This method should be able to detect cyclical definitions in the AuthorizationModel definition and return an appropriate error such as `ErrCyclicalModelDefinition` or something like that. There's no need to include this in a separate package. It's directly related to typesystem validations on authz models.

Reply:
Added the code inside `typesystem`